python/ray/tests/test_unavailable_actors.py

```diff
@@ -0,0 +1,184 @@
+
+        # Kill the actor process. The actor will restart so we get a temporal
+        # unavailable.
+        pid = ray.get(a.getpid.remote())
+        os.kill(pid, signal.SIGKILL)
+        with pytest.raises(ActorUnavailableError):
+            print(ray.get(a.slow_increment.remote(2, 0.1)))
+
+        # Actor restarting for 5s. In this period, we get a RESTARTING issue.
+        with pytest.raises(
+            ActorUnavailableError, match="The actor is in RESTARTING state"
+        ):
+            print(ray.get(a.slow_increment.remote(2, 0.1)))
+        time.sleep(6)
+
+        # After actor started, next calls are OK. However the previous actor instance's
+        # state is lost.
+        total = 0
+        for i in range(10):
+            total += i
+            assert ray.get(a.slow_increment.remote(i, 0.1)) == total
+
+        # Kill the actor again. This time it's not gonna restart so RayActorError.
+        pid = ray.get(a.getpid.remote())
+        os.kill(pid, signal.SIGKILL)
+        with pytest.raises(RayActorError):
+            print(ray.get(a.slow_increment.remote(1, 0.1)))
+
+    call_from(body, caller)
+
+
+@pytest.mark.parametrize(
+    "caller",
+    ["actor", "task", "driver"],
+)
+def test_actor_unavailable_norestart(ray_start_regular, caller):
+    def body():
+        a = Counter.remote()
+        assert ray.get(a.sum.remote(1, 2)) == 3
+        pid = ray.get(a.getpid.remote())
+
+        # Kill the actor process. The actor died permanently so RayActorError.
+        os.kill(pid, signal.SIGKILL)
+        with pytest.raises(RayActorError):
+            print(ray.get(a.sum.remote(1, 2)))
+
+    call_from(body, caller)
+
+
+@pytest.mark.skip(reason="https://github.com/ray-project/ray/issues/44357")
+@pytest.mark.parametrize(
+    "caller",
+    ["actor", "task", "driver"],
+)
+def test_generators_early_stop_unavailable(ray_start_regular, caller):
+    """
+    For a streaming generator, if a connection break happens, *some* elements may still
+    yield because they are received prior to the break. Then, two things can happen:
+    - an early StopIteration, if `next(gen)` is called.
+    - a ActorUnavailableError, if `ray.get(obj_ref)` is called.
+    """
+
+    def body():
+        a = Counter.remote()
+        pid = ray.get(a.getpid.remote())
+        total = 2000000
+        break_at = 5
+        gen = a.gen_iota.remote(total)
+        obj_refs = []
+        i = 0
+        for obj_ref in gen:
+            obj_refs.append(obj_ref)
+            if i == break_at:
+                print(f"breaking conns at {i}")
+                close_common_connections(pid)
+            i += 1
+        # StopIteration happened before `total` elements reached.
+        # On my laptop, 11120 elements are collected.
+        print(f"collected {len(obj_refs)} elements")
+        assert len(obj_refs) < total
+        with pytest.raises(ActorUnavailableError):
+            ray.get(obj_refs)
+
+    call_from(body, caller)
+
```

Comment:
How does actor method cancellation work if an actor is temporarily unavailable? If I do

```python
ray.cancel(actor_handle.run.remote())
```

is there a chance that the cancellation doesn't go through if the actor is unavailable? And does this also apply to recursive cancellations?

Reply:
if the actor is unavailable, the caller can not talk to the worker process. So:

- if the request has delivered to the actor -> the actor runs the method task, and caller can't cancel
- if the request has not yet delivered to the actor -> the actor did not run the method task, and there's nothing to cancel

so no, cancel does not work if the actor is unavailable.

Reply:
> if the request has delivered to the actor -> the actor runs the method task, and caller can't cancel

To confirm the behavior, under normal conditions, the caller can cancel the actor's method task while it's running (see `Actor Task` section in [API docs](https://docs.ray.io/en/latest/ray-core/api/doc/ray.cancel.html#ray-cancel)). However, when the actor is unavailable, the cancellation never makes it to the actor, so the method isn't cancelled. Is this correct?

A few questions about this:

1. What happens if `ray.cancel()` is called on a long running actor method while the actor is unavailable? When the actor is available again, will the cancellation go through automatically?
2. Will `ray.cancel()` raise any exception or warning when called on an unavailable actor, so the canceler can retry?
    * Is there any way to build these retries into Ray Core itself, so cancellations eventually succeed?

Reply:
@rynewang I think the expected behavior here would be that the underlying RPC to cancel the method would be retried until the actor becomes available (or is confirmed dead). Is there a technical blocker for why we can't do that?

cc @rkooo567 as I think he's most informed on cancellation codepath

Reply:
@rynewang this case still needs to be covered. Cancellation should not fail silently if the actor is unavailable.

If this increases the scope considerably, we can handle it in a separate PR. Either way though, we should prioritize this because it could affect disconnect behavior in Serve.

Reply:
For an unavailable actor, if we give a remote task, it will be delivered at most once and the actor may or may not receive it. If we give a cancellation, however, you mean we want to do *at least once* to make sure it's delivered. Do we have a criteria behind this difference?